blockdag/ghostdag.go

```diff
@@ -0,0 +1,132 @@
+			if anticoneSet.contains(parent) || past.contains(parent) {
+				continue
+			}
+			isAncestorOfSelectedParent, err := dag.isAncestorOf(parent, node.selectedParent)
+			if err != nil {
+				return nil, err
+			}
+			if isAncestorOfSelectedParent {
+				past.add(parent)
+				continue
+			}
+			anticoneSet.add(parent)
+			anticoneHeap.Push(parent)
+			queue = append(queue, parent)
+		}
+	}
+	return &anticoneHeap, nil
+}
+
+// blueAnticoneSize returns the blue anticone size of 'block' from the worldview of 'context'.
+// Expects 'block' to be ∈ blue-set(context)
+func (dag *BlockDAG) blueAnticoneSize(block, context *blockNode) (uint32, error) {
+	for node := context; node != nil; node = node.selectedParent {
+		if blueAnticoneSize, ok := node.bluesAnticoneSizes[*block.hash]; ok {
+			return blueAnticoneSize, nil
+		}
+	}
+	return 0, errors.Errorf("block %s is not in blue-set of %s", block.hash, context.hash)
+}
+
+func (dag *BlockDAG) ghostdag(newNode *blockNode) (*blockHeap, error) {
+	newNode.selectedParent = newNode.parents.bluest()
+	newNode.bluesAnticoneSizes[*newNode.hash] = 0
+	newNode.blues = append(newNode.blues, newNode.selectedParent)
+	selectedParentAnticone, err := dag.selectedParentAnticone(newNode)
+	if err != nil {
+		return nil, err
+	}
+
+	for selectedParentAnticone.Len() > 0 {
+		blueCandidate := selectedParentAnticone.pop()
+		candidateBluesAnticoneSizes := make(map[*blockNode]uint32)
+		var candidateAnticoneSize uint32
+		possiblyBlue := true
+
+		for chainBlock := newNode; possiblyBlue; chainBlock = chainBlock.selectedParent {
+			// If blueCandidate is in the future of chainBlock, it means
+			// that all remaining blues are in past(chainBlock) and thus
+			// in past(blueCandidate). In this case we know for sure that
+			// anticone(blueCandidate) will not exceed K, and we can mark
+			// it as blue.
+			//
+			// newNode is always in the future of blueCandidate, so there's
+			// no point in checking it.
+			if chainBlock != newNode {
+				if isAncestorOf, err := dag.isAncestorOf(chainBlock, blueCandidate); err != nil {
+					return nil, err
+				} else if isAncestorOf {
+					break
+				}
+			}
+
+			for _, block := range chainBlock.blues {
+				// Skip blocks that exists in the past of blueCandidate.
+				// We already checked it for chainBlock above, so if the
+				// block is chainBlock, there's no need to recheck.
+				if block != chainBlock {
+					if isAncestorOf, err := dag.isAncestorOf(block, blueCandidate); err != nil {
+						return nil, err
+					} else if isAncestorOf {
+						continue
+					}
+				}
+
+				candidateBluesAnticoneSizes[block], err = dag.blueAnticoneSize(block, newNode)
+				if err != nil {
+					return nil, err
+				}
+				candidateAnticoneSize++
+				if candidateAnticoneSize > dag.dagParams.K || candidateBluesAnticoneSizes[block] == dag.dagParams.K {
+					// Two possible k-cluster violations here:
+					// 	(i) The candidate blue anticone now became larger than k
```

Comment:
I'd split the if into two lines, 1 per condition, and near every condition place the comment that says what kind of violation it is.